python/tests/test_tables.py

```diff
@@ -2266,3 +2266,336 @@ def test_set_columns_not_implemented(self):
         t = tskit.BaseTable(None, None)
         with self.assertRaises(NotImplementedError):
             t.set_columns()
+
+
+class TestSubsetTables(unittest.TestCase):
+
+    seed = 12345
+
+    def subset_ragged_columns(self, packed, offset, subset, name):
+        """
+        Subsets a pair of ragged columns by the rows specified in `subset`.
+        :param numpy.ndarray packed: The flattened array of data.
+        :param numpy.ndarray offset: The array of offsets into the ``packed`` array.
+        : param str name: The name of the ragged columns.
+        :return: A tuple with the subsetted flattened array of data and offsets.
+        :rtype: tuple[numpy.ndarray]
+
+        """
+        if name in ["ancestral_state", "derived_state", "record", "timestamp"]:
+            unpack = tskit.unpack_strings
+            pack = tskit.pack_strings
+        elif name == "location":
+            unpack = tskit.unpack_arrays
+            pack = tskit.pack_arrays
+        else:
+            unpack = tskit.unpack_bytes
+            pack = tskit.pack_bytes
+        unpacked = unpack(packed, offset)
+        unpacked = [unpacked[i] for i in subset]
+        return pack(unpacked)
+
+    def np_subset(self, tables, nodes, record_provenance=True):
```

Comment:
I agree this naive version would be the best as a test, but note we have the `verify_subset` which is more obviously correct. There we go row by row for each table.

I don't think we need the `np_subset` in the test to make it robust, but we decided to keep it there in case we want to have a look at it in the future. Importantly, with the numpy version we discovered that even though the numpy version is a little easier to write, the C version is waay faster.

Reply:
> Importantly, with the numpy version we discovered that even though the numpy version is a little easier to write, the C version is waay faster.

This was a good exercise (for me at least) - naively, I thought that the numpy version would be easier to write and just as fast; in fact, it was harder to write and slower. (see the graph above) Maybe we could make it just as fast with additional work, but that would defeat the purpose. I lean towards leaving it in but would support removing it if you have a stronger opinion.

Reply:
Which graph is this @petrelharp? 

Sure, we can keep the numpy version in, but I'd still like to see the naive dumb-as-rocks Python row-by-row version. I agree that the verify function does this, but it's not quite the same thing to check something as it is to write it down and actually do it. We can always compare all three versions to each other. In terms of making sure that the ragged column stuff is done properly, I think this is really the best and simplest way to do it.

Reply:
Took me a minute to find, but [here](https://github.com/tskit-dev/tskit/pull/663#issuecomment-638340711) it is.

I'm not yet convinced that the naive version would make the testing more robust here, but I'll discuss with @mufernando.

Reply:
> Took me a minute to find, but here it is.

Thanks - that is an impressive difference. Writing non-simple things in numpy is hard!

> I'm not yet convinced that the naive version would make the testing more robust here, but I'll discuss with @mufernando.

It's not just about testing robustness @petrelharp, it's about clarity of what the test is actually *doing*. We need to be able to read these tests in 5 years time and understand how they work - hence, my push for a simple, direct, Python implementation.

We'd have written it in the time it takes to write these comments...